### ISSUE TITLE
Appearance of GTK modal dialogs not being correctly set

### ISSUE THREAD

**Issue by spl237** [labwc/labwc#1698]
Labwc's handling of dialogs which are set as modal in GTK applications is not completely consistent with their behaviour under X.

The Geany editor is a good demonstration, as it has both modal and non-modal dialogs.

Non-modal dialogs, such as the search dialog, behave identically in X and in labwc - they can be opened, they stay in front of the parent but can lose focus if the parent is clicked while remaining open (and the title bar and GTK foreground/background theme is updated accordingly), and keyboard and mouse input can be directed to either the parent or the dialog as expected.

In X, a modal dialog, such as the open file dialog, completely locks out access to the parent. It is not possible to focus the parent by clicking on it, or indeed to interact with it in any way, until the modal dialog is closed. This is, as far as I understand it, the expected behaviour of a modal dialog.

In labwc, it is possible to click away from the open file dialog back to the parent - the title bar colours update to suggest the parent now has focus, and it adopts the GTK theme for a foreground window rather than a background window, while the modal dialog adopts the background GTK theme. However, keyboard and mouse input are (correctly) blocked on the parent while the modal is open, so it appears to be the indication of focus and foreground / background which is incorrect.

I believe the only fix required is to make it such that clicking on the parent when a modal dialog is open does not update the appearance (titlebar and GTK foreground / background theme) of the parent or the dialog.

**Comment by Consolatis** [comment]
I don't think there is a modal flag in wayland yet (AFAIR there was some discussion about it in wayland-protocols but I didn't follow up on that).

As such we decided to go the current route because otherwise non-modal dialogs would not be possible at all.

**Comment by spl237** [comment]
Ah, ok - so effectively you are treating modal and non-modal dialogs the same as far as titlebar and theme go, the only difference is the parent blocking input on a modal? Is that correct?

**Comment by Consolatis** [comment]
Yes, correct from what I remember without looking through the sources.

**Comment by spl237** [comment]
Just thinking out loud - the change to foreground / background theme colours must be managed by GTK (as it is the GTK widgets which get their colours changed). So something (GTK?) knows to lock out interaction with a parent while a modal is on display, but not to lock its theme to background. I wonder if this might perhaps be fixable within GTK itself even if it cannot be fixed in Wayland?

Although I'm not sure how that would work with the title bar...

**Comment by Consolatis** [comment]
That would only work for CSD applications, they could theoretically ignore the set_activated call (sent by labwc when changing keyboard focus) when they have a modal dialog open. Having a proper modal flag for wayland would be the better solution though.

**Comment by Consolatis** [comment]
Should now work in the master branch, requires support from the client as well though. Gtk4 seems to support it, Gtk3 apparently does not.
